src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala

```diff
@@ -0,0 +1,133 @@
+      position: Long) {
+  }
+
+  case class GenomeRange(
+    contig: String,
+    start: Long,
+    stop: Long)
+
+  object Caller extends SparkCommand[Arguments] {
+    override val name = "structural-variant"
+    override val description = "Find structural variants, i.e. large insertions or deletions, inversions and translocations."
+
+    // The sign of inferredInsertSize follows the orientation of the read. This returns
+    // an insert size which is positive in the common case, regardless of the orientation.
+    def orientedInsertSize(r: PairedRead[MappedRead]): Option[Int] = {
+      val sgn = if (r.isPositiveStrand) { 1 } else { -1 }
+      r.mateAlignmentProperties.flatMap(_.inferredInsertSize).map(x => x * sgn)
+    }
+
+    // Coalesce sequences of values separated by blockSize into single ranges.
+    // The sequence of values must be sorted.
+    def coalesceAdjacent(xs: IndexedSeq[Long], blockSize: Long): Iterator[(Long, Long)] = {
+      val rangeStarts = ArrayBuffer(0)
+      for (i <- 1 until xs.length) {
+        if (xs(i - 1) + blockSize != xs(i)) {
+          rangeStarts += i
+        }
+      }
+      rangeStarts += xs.length
+
+      rangeStarts.sliding(2).map(items => {
+        val startIdx = items(0)
+        val nextStartIdx = items(1)
+        (xs(startIdx), xs(nextStartIdx - 1))
+      })
+    }
+
+    // Returns a sequence of all blocks which overlap the read, its mate or the insert between them.
+    def matedReadBlocks(r: PairedRead[MappedRead]): Seq[Long] = {
+      if (!r.isMateMapped) {
+        Seq[Long]()
+      } else {
+        val mateStart = r.mateAlignmentProperties.map(mp => mp.start).getOrElse(r.read.start)
+        val start = List(r.read.start, r.read.end, mateStart).min
+        val roundedStart = (1.0 * start / BLOCK_SIZE).toLong * BLOCK_SIZE
+        val insertSize = orientedInsertSize(r).getOrElse(0)
+        (roundedStart to (roundedStart + insertSize) by BLOCK_SIZE)
+      }
+    }
+
+    override def run(args: Arguments, sc: SparkContext): Unit = {
+      val readSet = Common.loadReadsFromArguments(args, sc, Read.InputFilters(nonDuplicate = true))
+      val pairedMappedReads = readSet.mappedPairedReads
+      val firstInPair = pairedMappedReads.filter(_.isFirstInPair)
+
+      // Pare down to mate pairs which aren't highly displaced & aren't inverted.
+      val readsInRange = firstInPair.filter(r => {
+        val insertSize = orientedInsertSize(r).getOrElse(0)
+        (insertSize > 0) && (insertSize < MAX_INSERT_SIZE)
+      })
+      readsInRange.persist()
+
+      val insertSizes = readsInRange.flatMap(orientedInsertSize)
+      val insertStats = insertSizes.stats()
+      println("Stats on inferredInsertSize: " + insertStats)
+
+      // Create a map from (block) -> [insert sizes]
+      // TODO: take greater advantage of data locality than .groupByKey() does?
+      val insertsByBlock = readsInRange.flatMap(r => {
+        val insertSize = orientedInsertSize(r).getOrElse(0)
```

Comment:
```
(for {
  r <- readsInRange
  pos <- matedReadBlocks(r)
  insertSize <- orientedInsertSize(r)
} yield {
  Location(r.read.referenceContig, pos) -> insertSize
}).groupByKey
```

Tricky thing about `for`s and chained `flatMap`s and `map`s in Scala: you can't `flatMap` over an `Option` then a list; e.g. this would give a not-very-descriptive error about inferred type something somethings:
```
  r <- readsInRange
  insertSize <- orientedInsertSize(r)
  pos <- matedReadBlocks(r)
```